nucliadb/nucliadb/common/cluster/rollover.py

```diff
@@ -0,0 +1,353 @@
+#
+import argparse
+import asyncio
+import logging
+from datetime import datetime
+from typing import Optional
+
+from nucliadb.common.cluster import manager as cluster_manager
+from nucliadb.common.context import ApplicationContext
+from nucliadb.common.datamanagers.cluster import ClusterDataManager
+from nucliadb.common.datamanagers.resources import ResourcesDataManager
+from nucliadb_protos import writer_pb2
+from nucliadb_telemetry import errors
+from nucliadb_utils import const
+
+from .exceptions import ExhaustedNodesError
+from .manager import get_index_node
+from .settings import settings
+
+logger = logging.getLogger(__name__)
+
+
+class UnexpectedRolloverError(Exception):
+    pass
+
+
+async def create_rollover_shards(
+    app_context: ApplicationContext, kbid: str
+) -> writer_pb2.Shards:
+    logger.warning("Creating rollover shards", extra={"kbid": kbid})
+    sm = app_context.shard_manager
+    cluster_datamanager = ClusterDataManager(app_context.kv_driver)
+
+    existing_rollover_shards = await cluster_datamanager.get_kb_rollover_shards(kbid)
+    if existing_rollover_shards is not None:
+        logger.warning(
+            "Rollover shards already exist, deleting them and restarting",
+            extra={"kbid": kbid},
+        )
+        for shard in existing_rollover_shards.shards:
+            await sm.rollback_shard(shard)
+
+    kb_shards = await cluster_datamanager.get_kb_shards(kbid)
+    if kb_shards is None:
+        raise UnexpectedRolloverError(f"No shards found for KB {kbid}")
+
+    # create new shards
+    created_shards = []
+    try:
+        for shard in kb_shards.shards:
+            nodes = cluster_manager.sorted_nodes()
+            shard.ClearField("replicas")
+            # Attempt to create configured number of replicas
+            replicas_created = 0
+            while replicas_created < settings.node_replicas:
+                try:
+                    node_id = nodes.pop(0)
+                except IndexError:
+                    # It was not possible to find enough nodes
+                    # available/responsive to create the required replicas
+                    raise ExhaustedNodesError()
+
+                node = get_index_node(node_id)
+                if node is None:
+                    logger.error(f"Node {node_id} is not found or not available")
+                    continue
+                try:
+                    shard_created = await node.new_shard(
+                        kbid, similarity=kb_shards.similarity
+                    )
+                except Exception as e:
+                    errors.capture_exception(e)
+                    logger.error(f"Error creating new shard at {node}")
+                    continue
+
+                replica = writer_pb2.ShardReplica(node=str(node_id))
+                replica.shard.CopyFrom(shard_created)
+                shard.replicas.append(replica)
+                created_shards.append(shard)
+                replicas_created += 1
```

Comment:
Maybe not for this PR, but as it is kind of the same code we have in `create_shard_by_kbid` , this could be refactored into a util / function?

Reply:
I actually tried at some point but the logic is a bit wonky and I wasn't feeling good about the generic abstraction